MaterialLibrary/src/main/java/io/doist/material/elevation/ElevationWrapperDrawable.java

```diff
@@ -0,0 +1,488 @@
+     *
+     * @param force forces the update to go through, even if there's no indication that the shadow has changed.
+     * @param async runs the update in the background instead of the calling thread.
+     */
+    private void update(boolean force, boolean async) {
+        View view = mViewRef.get();
+        if (view != null) {
+            view.getLocationOnScreen(mScreenLocation);
+            int left = mScreenLocation[0] + getPaddingLeft();
+            int top = mScreenLocation[1] + getPaddingTop();
+            int right = left + mBounds.width();
+            int bottom = top + mBounds.height();
+
+            if (force || left != mLeft || top != mTop || right != mRight || bottom != mBottom) {
+                mLeft = left;
+                mTop = top;
+                mRight = right;
+                mBottom = bottom;
+
+                int shadowLengthLeft = getShadowLengthLeft(left);
+                int shadowLengthTop = getShadowLengthTop();
+                int shadowLengthRight = getShadowLengthRight(right);
+                int shadowLengthBottom = getShadowLengthBottom(bottom);
+
+                boolean leftDirty = mShowShadowLeft && (force || shadowLengthLeft != mShadowLengthLeft);
+                boolean topDirty = mShowShadowTop && (force || shadowLengthTop != mShadowLengthTop);
+                boolean rightDirty = mShowShadowRight && (force || shadowLengthRight != mShadowLengthRight);
+                boolean bottomDirty = mShowShadowBottom && (force || shadowLengthBottom != mShadowLengthBottom);
+
+                if (leftDirty || topDirty || rightDirty || bottomDirty) {
+                    float shadowAlphaLeft = shadowLengthLeft > shadowLengthRight ? SIDE_ALPHA : AMBIENT_ALPHA;
+                    float shadowAlphaTop = AMBIENT_ALPHA;
+                    float shadowAlphaRight = shadowLengthRight > shadowLengthLeft ? SIDE_ALPHA : AMBIENT_ALPHA;
+                    float shadowAlphaBottom =
+                            MIN_BOTTOM_ALPHA + INC_BOTTOM_ALPHA * shadowLengthBottom / getPaddingBottom();
+
+                    ElevationUpdateRunnable runnable =
+                            new ElevationUpdateRunnable(
+                                    mLeft, mTop, mRight, mBottom, mCornerRadius,
+                                    shadowLengthLeft, shadowLengthTop, shadowLengthRight, shadowLengthBottom,
+                                    shadowAlphaLeft, shadowAlphaTop, shadowAlphaRight, shadowAlphaBottom,
+                                    leftDirty, topDirty, rightDirty, bottomDirty, this);
+                    if (async) {
+                        sShadowExecutor.execute(runnable);
+                    } else {
+                        runnable.run();
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Returns the ambient shadow for the current elevation.
+     */
+    private int getShadowLengthAmbient() {
+        return (int) Math.max(1, mElevation * 3 / 8);
```

Comment:
Ambient shadow should always exist (at least according to my tests), as long as the elevation is `> 0` (this method won't be called otherwise). 

A better version would probably be: `Math.ceil(mElevation * 3 / 8)`.